### ISSUE TITLE
New various results.

### ISSUE THREAD

**Issue by iperov** [deepfakes/faceswap-playground#131]
(sry bad english)

I keep working on new OpenDeepFaceSwap every day. I will share it of course.

I choosed worst case data where source face is brighter, dest face is darker, skin tone differents. Source face has bright blue background which is BAD for original faceswap and fakeapp.

All models targets same epoch (~4-5h training) with same batch size. 

First look at work of 'super improved DF full face match warper':

![warper working](https://user-images.githubusercontent.com/8076202/38776936-7b929cb6-40b0-11e8-8605-4fba48885c69.gif)

Original 128 with super improved DF full face match warper:

![012090 f128](https://user-images.githubusercontent.com/8076202/38776939-9cf5bd02-40b0-11e8-9ff2-aa10ef680c54.jpg)
![out00002 f128](https://user-images.githubusercontent.com/8076202/38776940-a3420116-40b0-11e8-9668-b4464d706e8c.png)
![out00168 f128](https://user-images.githubusercontent.com/8076202/38776941-a4c53e54-40b0-11e8-99ae-6e7823e6a18c.png)

Original 128 with DSSIM loss func (DF model) with super improved DF full face match warper:

![012090 df](https://user-images.githubusercontent.com/8076202/38776945-af6431b2-40b0-11e8-9c2d-f22a3b3e724b.jpg)
![out00002 df](https://user-images.githubusercontent.com/8076202/38776946-b07f2bd8-40b0-11e8-956d-673ca5057ed0.png)
![out00168 df](https://user-images.githubusercontent.com/8076202/38776948-b12a498c-40b0-11e8-959c-66e9c65eab52.png)

^ better than original. 

DSSIM loss makes result better than without it, also it suppress of learning background, so in theory more neurons will learn face instead of background.

Also we can see original model just averages source faces, therefore source face has half closed eyes.

IAE 128 with DSSIM loss func with super improved DF full face match warper:

![012090 iae original](https://user-images.githubusercontent.com/8076202/38776968-013042b0-40b1-11e8-8b22-e37d96c47d81.jpg)
![out00002 iae original](https://user-images.githubusercontent.com/8076202/38776969-01f0a08c-40b1-11e8-86c0-ae14f6270bf5.png)
![out00168 iae original](https://user-images.githubusercontent.com/8076202/38776970-02c9003a-40b1-11e8-98e3-0399c6d7a258.png)

^ bettter than DF, face more properly aligned, eyes opened and looking straight, while all facial features saved.

IAE 128 experimental with DSSIM loss func with super improved DF full face match warper:

![012090 iae4 8](https://user-images.githubusercontent.com/8076202/38776971-0450cf32-40b1-11e8-8bee-50aa3876593d.jpg)
![out00002 iae 4 8](https://user-images.githubusercontent.com/8076202/38776972-052bff6c-40b1-11e8-9bb3-3973d75ba410.png)
![out00168 iae 4 8](https://user-images.githubusercontent.com/8076202/38776973-05c3475a-40b1-11e8-918f-92dd0207cac9.png)

Compare in detail IAE128 and IAE128 experimental:
![1](https://user-images.githubusercontent.com/8076202/38776996-51d8e60e-40b1-11e8-8fe8-d2cb469ab180.gif)

In IAE128 experimental we can see better supressing of blue background of source and better skin fade matching.

In IAE128 experimental I found new way how to make IAE better.
Main idea is gather NOT only 4x4 features in intermediate layers, but also 8x8.
```
def Intermediate(self, is_low_mem):
    input_layer = self.keras.layers.Input(shape=(None, 4 * 4 * 1024))

    neck = self.keras.layers.Dense(1024)(input_layer)
    
    x4x4 = self.keras.layers.Dense(4 * 4 * 512)(neck)
    x4x4 = self.keras.layers.Reshape((4, 4, 512))(x4x4)

    x8x8 = self.keras.layers.Dense(8 * 8 * 256)(neck)
    x8x8 = self.keras.layers.Reshape((8, 8, 256))(x8x8)
```

and then add it in decoder:
```
x = upscale(self.keras, x4x4, 512)
x = res(self.keras, x, 512)

x = self.keras.layers.Add()([x, x8x8]) 
```

but it consumes more VRAM. I would like to add 16x16 features but I cant due to lack of vram 6GB.

IAE training history https://webm.video/i/mX5TCJ.mp4

**Comment by iperov** [comment]
@dfaker, in last mp4 video can you say something about flickering holes in masks ?

**Comment by iperov** [comment]
IAE128 experimental 2:
![2018-04-15_19-37-03](https://user-images.githubusercontent.com/8076202/38780272-e7f3ccde-40e4-11e8-8c14-cc4fb1b8deb1.png)
![out00168 iae2](https://user-images.githubusercontent.com/8076202/38780306-9b264282-40e5-11e8-9e4e-f0862474930c.png)

better than IAE128 experimental 1 !!

**Comment by dfaker** [comment]
@iperov It seems to be identical in both axes for both the mask and the rgb. 
That'd lead me to guess that it's due to the network learning some bias in the averaged transformations of all of the samples in the last trained batch.

**Comment by iperov** [comment]
@dfaker any suggestions how to fix it ?

**Comment by dfaker** [comment]
That's the sort of thing that lead me to put the mask and rgb layers into separate generators, I saw both this (an 'edge detection' on edge like features in the face causing voids in the mask) and the mask layer bleeding thorough to to add a color cast or lighting effect to the rgb face data.

Sadly my best solution was, as you saw, to split them into two parallel decoders.

**Comment by iperov** [comment]
@dfaker um... Actually this mask flickering bug in your model. I use same separated mask trainer.

```
def Decoder(self, is_low_mem):
            
        input  = self.keras.layers.Input(shape=(4, 4, 1024))
        x = input
        x = upscale(self.keras, x, 512)
        x = res(self.keras, x, 512)        
        
        x = upscale(self.keras, x, 256)
        x = res(self.keras, x, 256)        
        
        x = upscale(self.keras, x, 128)
        x = res(self.keras, x, 128)        
        
        x = upscale(self.keras, x, 64)
        x = res(self.keras, x, 64)        
        
        x = upscale(self.keras, x, 32)
        x = res(self.keras, x, 32)

        x = self.keras.layers.convolutional.Conv2D(3, kernel_size=5, padding='same', activation='sigmoid')(x)
        
        y = input  #mask decoder
        y = upscale(self.keras, y, 256)
        y = upscale(self.keras, y, 128)
        y = upscale(self.keras, y, 64)
        y = upscale(self.keras, y, 32)
        y = upscale(self.keras, y, 16)
        y = self.keras.layers.convolutional.Conv2D( 1, kernel_size=5, padding='same', activation='sigmoid' )(y)

        return self.keras.models.Model( [input], [x,y])
```

**Comment by iperov** [comment]
looks like it depends on dimensions of upscale layers.

Finally IAE experimental 3 ! Best model of all time.

this video demonstrates result:
F128 -> DF -> IAE original -> IAE 1 -> IAE 2 -> IAE 3

https://webm.video/i/n4ddXk.mp4

**Comment by iperov** [comment]
Total size of .h5 files of IAE 3 is 792MB.

**Comment by iperov** [comment]
@acsaga thanks for your original IAE.

**Comment by dfaker** [comment]
@iperov I was under the impression that you had made some innovations in the network itself, not just re-using the existing model. Perhaps your tweaks are all in the encoder not just the sample generation.

Are you suggesting that the areas of dark within the mask of your failed cases are some kind of deconvolution artifact? They don't at all have the classic look one would expect.

**Comment by iperov** [comment]
> Perhaps your tweaks are all in the encoder not just the sample generation.
yeah
1) your match warper
2) smart sample feeder by sorted yaw to prevent model overfit with only one yaw.
3) your dssim loss, because original IAE without dssim 
4) in IAE3:

```
    def Encoder(self, input_layer, is_low_mem):
        x64 = input_layer
        x32 = conv(self.keras, x64, 128)
        x16 = conv(self.keras, x32, 256)
        x8 = conv(self.keras, x16, 512)
        x4 = conv(self.keras, x8, 1024)
        
        return self.keras.models.Model(input_layer, 
                [  
                    self.keras.layers.Flatten()(x32),
                    self.keras.layers.Flatten()(x16),
                    self.keras.layers.Flatten()(x8),
                    self.keras.layers.Flatten()(x4)
                ]
            )

    def Intermediate(self, is_low_mem):
        input_x32 = self.keras.layers.Input(shape=(None, 32 * 32 * 128))
        input_x16 = self.keras.layers.Input(shape=(None, 16 * 16 * 256))
        input_x8 = self.keras.layers.Input(shape=(None, 8 * 8 * 512))
        input_x4 = self.keras.layers.Input(shape=(None, 4 * 4 * 1024))
        
        x32 = self.keras.layers.Dense(64)(input_x32)
        x16 = self.keras.layers.Dense(128)(input_x16)
        x8 = self.keras.layers.Dense(256)(input_x8)
        x4 = self.keras.layers.Dense(512)(input_x4)
        
        x = self.keras.layers.Concatenate()( [x32, x16, x8, x4] )
        
        x = self.keras.layers.Dense(4 * 4 * 512)(x)
        x = self.keras.layers.Reshape((4, 4, 512))(x)
            
        return self.keras.models.Model( [input_x32, input_x16,  input_x8, input_x4], [x]  )
```
and it can be upgraded when new cards with more VRAM will appear in future.
This models works on 6GB with only 2 batch size lool.

> Are you suggesting that the areas of dark within the mask of your failed cases are some kind of deconvolution artifact?

problem almost fixed when I increased dims to 512 from 256.
```
        y = upscale(self.keras, y, 512)
        y = upscale(self.keras, y, 256)
        y = upscale(self.keras, y, 128)
        y = upscale(self.keras, y, 64)
        y = upscale(self.keras, y, 32)
```
now much better.

https://webm.video/i/Xb3qDs.mp4
